### PR TITLE
golang versiond ovn-kube-gateway-helper

### PR DESCRIPTION
**Pull request by YaoZengzeng**

### CODE REVIEW

**Comment by rajatchopra** [comment]
Will test and evaluate this. Eventually I feel we should have at most two executables. One for cni, all others should be part of one binary with different options.

**Comment by YaoZengzeng** [comment]
@rajatchopra agree. Two binaries will be good.

**Comment by shettyg** [comment]
The python code does "sync_services", which the golang code need not do in between its run because of the inbuilt cache. 

But consider the case where, the gateway helper has added a few entries based on a ADD nodeport event. The gateway helper is stopped. Now the service is deleted in k8s. Some time passes by. You start gateway helper. It will not get the delete event and hence it will not delete the old flows.  So I think you will need a "sync" after registering with k8s service API. You should also wait processing any events till the sync is complete.

**Comment by YaoZengzeng** [comment]
@shettyg Updated

**Comment by rajatchopra** [comment]
The watch should begin after the sync is complete. Also it should start from the resource version of the latest sync.

**Comment by rajatchopra** [comment]
> The watch should begin after the sync is complete. Also it should start from the resource version of the latest sync.
@YaoZengzeng I see that you have handled the sync for existing items so that delete can be managed. I am updating the watch function so that a processExisting function can be accommodated with all watches. Hold off on this PR while I get that going.
Thanks.

File: go-controller/cmd/ovn-kube-gateway-helper/ovn-kube-gateway-helper.go [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+			if err != nil {
+				logrus.Errorf("del flow of %s failed: %q", physicalBridge, stdout)
+			}
+		}
+	}
+
+	return nil
+}
+
+func main() {
```
Comment (rajatchopra) [inline go-controller/cmd/ovn-kube-gateway-helper/ovn-kube-gateway-helper.go]: Please break this function up. The main function is too long. Handle the parse args part separately from the logic maybe.

File: go-controller/cmd/ovn-kube-gateway-helper/ovn-kube-gateway-helper.go [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+
+	StartServiceWatch(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			service, ok := obj.(*kapi.Service)
+			if !ok {
+				logrus.Errorf("couldn't get service from obj")
+				return
+			}
+
+			if service.Spec.Type != kapi.ServiceTypeNodePort || len(service.Spec.Ports) == 0 {
```
Comment (rajatchopra) [inline go-controller/cmd/ovn-kube-gateway-helper/ovn-kube-gateway-helper.go]: Move the callback logic to a separate function.

File: go-controller/cmd/ovn-kube-gateway-helper/ovn-kube-gateway-helper.go [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+				return
+			}
+
+			for _, svcPort := range service.Spec.Ports {
```
Comment (rajatchopra) [inline go-controller/cmd/ovn-kube-gateway-helper/ovn-kube-gateway-helper.go]: Same here. Please call this in another function.

**Comment by shettyg** [comment]
I have taken some code from this pull request, added you as a co-author and sent in a new pull request here: https://github.com/openvswitch/ovn-kubernetes/pull/258

**Pull request closed without merging**